### PR TITLE
feat: ingest performance, job coordination, memory export

### PR DESCRIPTION
**Pull request by equationalapplications**
## Summary

Implements the full spec from `docs/superpowers/specs/2026-04-30-ingest-perf-and-export.md`.

### Changes

**Performance**
- `chunkText` helper replaces O(n²) regex splitter — paragraph-first, sentence-second, whitespace-third, hard-cut fallback
- `ingestDocument` parallelises LLM calls across chunks via `Promise.all` (3–5× wall-clock speedup on multi-chunk docs)
- Cross-chunk fact dedup by normalised title (first-wins) before single transactional DB write
- `maxChunkLength` default raised 6000 → 12000; new `chunkOverlap` config (default 400)

**Job coordination**
- `runLibrarian` and `runHeal` now use separate mutex keys (`…:librarian` / `…:heal`) so they no longer block each other
- `getEntityStatus(entityId)` — synchronous snapshot of in-flight jobs, safe to call from React render

**Prompt budgets**
- `INGEST_SYSTEM_PROMPT` and `LIBRARIAN_SYSTEM_PROMPT` body budget raised 200 → 800 chars

**Export / Import**
- `WikiMemory.exportDump(entityIds?)` — export memory as `MemoryDump`
- `WikiMemory.importDump(dump, { merge? })` — restore from dump; `merge:false` (default) replaces, `merge:true` skips existing ids; events always appended
- `formatMemoryDump(dump)` — pure helper: JSON manifest + one `.md` per entity
- `useWikiExport()` React hook

**Error handling**
- New `WikiBusyError` class with typed `operation` and `entityId` fields

**Tests**
- 35 tests across 7 files (Vitest, added from zero)

---

> ⚠️ **BREAKING CHANGE:** `runLibrarian()` and `runHeal()` now throw `WikiBusyError` when a run is already active for that entity and operation, instead of silently returning.

### CODE REVIEW

File: src/WikiMemory.ts [hunk 1]
```diff
@@ -568,74 +799,81 @@ export class WikiMemory {
...
     const sourceHash = normalizeSourceHash(params.sourceHash);
     if (!sourceHash) throw new Error('Invalid sourceHash (must be 64-char hex string)');
 
-    const maxChunkLength = params.maxChunkLength ?? this.options.config?.maxChunkLength ?? 6000;
-    if (!Number.isInteger(maxChunkLength) || maxChunkLength < 2) {
-      throw new Error('maxChunkLength must be an integer greater than or equal to 2');
-    }
-    
+    const maxChunkLength = params.maxChunkLength ?? this.options.config?.maxChunkLength ?? 12000;
+    const chunkOverlap = params.chunkOverlap ?? this.options.config?.chunkOverlap ?? 400;
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts]: The developer should be able to configure the chunkOverlap. It should be added to:
```
export interface WikiConfig {
  tablePrefix?: string;               // DB table prefix, set at construction only
  maxFtsResults?: number;             // max results from full-text search queries
  pruneEventsAfter?: number;          // max events to keep per entity before pruning
  autoLibrarianThreshold?: number;    // event count that triggers auto-librarian
  autoHealThreshold?: number;         // fact count that triggers auto-heal
  orphanAfterDays?: number | null;    // days before an inferred fact with no access is marked stale
  staleInferredAfterDays?: number | null; // days before a stale fact is eligible for deletion
  maxChunkLength?: number;            // per-ingest chunk size (also overridable per-call)
}
```
and fully implemented.

File: package.json [hunk 2]
```diff
@@ -54,11 +55,13 @@
...
     "@types/react": "^19.0.0",
+    "@vitest/coverage-v8": "^4.1.5",
     "expo-sqlite": "^55.0.15",
     "react": "19.2.0",
     "semantic-release": "^24.0.0",
     "tsup": "^8.0.0",
-    "typescript": "^5.0.0"
+    "typescript": "^5.0.0",
+    "vitest": "^4.1.5"
   },
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply package.json:65]: expanding the peer range to include the Expo 55

File: src/WikiMemory.ts [hunk 3]
```diff
@@ -88,6 +88,78 @@ function safeSlice(value: string, start: number, end?: number): string {
   return value.slice(safeStart, safeEnd);
 }
 
+export function chunkText(
+  input: string,
+  maxChunkLength: number,
+  overlap: number
+): { chunks: string[]; truncated: boolean } {
+  const text = input.trim();
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts]: no need to export chunkText

**Comment by equationalapplications** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/equationalapplications/expo-llm-wiki/pull/4#pullrequestreview-4205507640)

**Comment by equationalapplications** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/equationalapplications/expo-llm-wiki/pull/4#pullrequestreview-4205973223)

File: src/WikiMemory.ts [hunk 4]
```diff
@@ -506,6 +616,160 @@ export class WikiMemory {
...
+        )
+      `);
+      ids = rows.map(r => r.entity_id);
+    }
+
+    const entities = Object.fromEntries(
+      await Promise.all(
+        ids.map(async (id): Promise<[string, MemoryBundle]> => [id, await this._getFullBundle(id)])
+      )
+    ) as Record<string, MemoryBundle>;
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts]: limiit the concurrency to something appropriate for a mobile device

**Comment by equationalapplications** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/equationalapplications/expo-llm-wiki/pull/4#pullrequestreview-4206647499)

File: src/WikiMemory.ts [hunk 5]
```diff
@@ -568,74 +875,84 @@ export class WikiMemory {
...
+          const responseText = await this.options.llmProvider.generateText({
+            systemPrompt: INGEST_SYSTEM_PROMPT,
+            userPrompt,
+          });
+          const result = parseJsonResponse<{ facts: ExtractedFact[] }>(responseText);
+          return (Array.isArray(result.facts) ? result.facts : [])
+            .map(validateFact)
+            .filter((f): f is ExtractedFact => f !== null);
+        })
+      );
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts]: Let the developer configure the concurrency and fall back to 1

**Comment by equationalapplications** [comment]
Unbounded concurrency — Valid. User confirms: default to 1, configurable. Need WikiConfig.chunkConcurrency, in-house worker-pool limiter (no new deps), update test that asserts max > 1 to pass an explicit concurrency value.

Mock event uniqueness key — Valid. Real schema has id as sole PRIMARY KEY; the mock checks (id, entity_id) which is wrong.

HEAL_SYSTEM_PROMPT body annotation — Valid. The spec said bump to 800, but the prompt's newFacts.body schema still reads "body": "string" with no limit. Add (max 800 chars) to match ingest/librarian.

Spec filename description — Valid. formatMemoryDump sanitizes IDs and may append a hash suffix; spec says ${entityId}.md.

useWikiExport spec signature — Valid. Spec shows {exportDump, isExporting} but shipped API is {execute, lastResult, isPending, error}.

fact ID chunking in importDump — Valid. Tasks already chunk to 500; facts query is unbounded. Fix to match.

**Comment by equationalapplications** [comment]
Findings

Medium: withConcurrency keeps scheduling work after the first chunk failure.
In WikiMemory.ts:163-174, if one worker rejects, Promise.all(concurrentWorkers) rejects immediately, but the other worker loops keep running and can continue pulling new chunk tasks. Then ingestDocument reaches its finally and clears activeIngestJobs while those background LLM calls may still be in flight. That undermines the busy guard and can still burn provider quota after the call has failed.
Fix: track firstError, stop scheduling new tasks once set, wait for all workers to settle, then rethrow. That preserves “no partial writes” and keeps the ingest job active until in-flight work has actually stopped.

Medium: chunkConcurrency is documented as a positive integer but not normalized or validated.
WikiMemory.ts:906 accepts the raw value from params/config, then WikiMemory.ts:163-174 uses it to size Array.from. 0 and negative values silently become 1; fractional values are effectively truncated by array-length coercion; NaN can create zero workers and return a sparse results array, causing a later TypeError when flattening results.
Since you want fallback to 1, I’d make that explicit with a small normalizer, e.g. “positive integer or fallback to 1,” and add tests for 0, negative, fractional, and NaN.

Low: the plan doc still contains stale implementation snippets that contradict the new spec.
The spec is updated, but the plan still shows Promise.all(chunks.map(...)) at 2026-04-30-ingest-perf-and-export.md:383, says to leave HEAL_SYSTEM_PROMPT unbounded at 2026-04-30-ingest-perf-and-export.md:489, and still has the older useWikiExport shape around 2026-04-30-ingest-perf-and-export.md:1209-1226. Since that file is explicitly for agentic workers, stale snippets are likely to cause future drift.

The event mock uniqueness change, HEAL prompt annotation, and fact-ID chunking all look sound. The test/typecheck sweep you asked for earlier was clean; lint remains unavailable because the repo has no lint script/config.

**Comment by equationalapplications** [comment]
Finding 1 — withConcurrency early-exit on failure: Workers now check a failed flag before picking new tasks and return immediately on error. Uses Promise.allSettled to wait for all in-flight calls to settle before rethrowing, so activeIngestJobs is cleared only after everything has actually resolved.

Finding 2 — chunkConcurrency normalization: NaN, 0, negative, and out-of-range values now fall back to 1 via Number.isFinite(raw) && raw >= 1 ? Math.floor(raw) : 1. Four new parameterized tests cover 0, -1, NaN, and a fractional value (1.7 → 1).

Finding 3 — Plan doc stale snippets: Updated all three locations: Promise.all → withConcurrency, HEAL prompt note corrected to say all three prompts were annotated, and useWikiExport hook shape updated to { execute, lastResult, isPending, error }.

**Comment by equationalapplications** [comment]
Both review comments addressed:

[getEntityStatus](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) / _getActiveIngestCounts — Removed _getActiveIngestCounts() entirely and replaced the [lastIndexOf(':')](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) key-parsing with a direct [startsWith(](http://_vscodecontentref_/8)

this.prefix:{entityId}:`)scan. This is correct even when [entityId](http://_vscodecontentref_/9) contains:, because [sourceRef](http://_vscodecontentref_/10) is normalized to exclude :so the trailing:` before [sourceRef](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is always the definitive separator.

[exportDump](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) ordering — Added [ORDER BY entity_id](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) to the subquery so the resulting entity list (and therefore [dump.entities](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) insertion order) is stable across runs.

**Comment by equationalapplications** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/equationalapplications/expo-llm-wiki/pull/4#pullrequestreview-4208269035)

File: src/WikiMemory.ts [hunk 6]
```diff
@@ -274,6 +375,28 @@ export class WikiMemory {
     return { facts, tasks, events: events.reverse() };
   }
 
+  async getMemoryBundle(entityId: string): Promise<{ facts: WikiFact[]; tasks: WikiTask[]; events: WikiEvent[] }> {
+    const [factsRaw, tasks, events] = await Promise.all([
+      this.db.getAllAsync<WikiFact>(
+        `SELECT * FROM ${this.prefix}entries WHERE entity_id = ? AND deleted_at IS NULL ORDER BY updated_at DESC`,
+        [entityId]
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts]: @copilot 
Yes to: `returning Promise<MemoryBundle> and refactoring to share the common DB fetch + tag JSON parsing so future schema/query changes don’t need to be updated in multiple places.`

File: src/WikiMemory.ts [hunk 7]
```diff
@@ -568,74 +882,89 @@ export class WikiMemory {
...
-    
+    const maxChunkLength = params.maxChunkLength ?? this.options.config?.maxChunkLength ?? 12000;
+    const chunkOverlap = Math.min(
+      params.chunkOverlap ?? this.options.config?.chunkOverlap ?? 400,
+      maxChunkLength - 1
+    );
+    const rawConcurrency = params.chunkConcurrency ?? this.options.config?.chunkConcurrency ?? 1;
+    const chunkConcurrency = Number.isFinite(rawConcurrency) && rawConcurrency >= 1
+      ? Math.floor(rawConcurrency)
+      : 1;
```
[the remark this answers is not in the document]
Comment (equationalapplications) [reply src/WikiMemory.ts:887]: @copilot apply changes based on [this feedback](https://github.com/equationalapplications/expo-llm-wiki/pull/4#discussion_r3170743438)

**Comment by equationalapplications** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/equationalapplications/expo-llm-wiki/pull/4#pullrequestreview-4208572121)
